### PR TITLE
vSphere Virtual Volume Support

### PR DESCRIPTION
**Pull request by gaurav-nelson**
WIP

### CODE REVIEW

**Comment by gaurav-nelson** [comment]
[Document: Backport vsphere fixes to openshift](https://trello.com/c/gSsay1n6/464-document-backport-vsphere-fixes-to-openshift)

File: architecture/additional_concepts/storage.adoc [hunk 1]
```diff
@@ -474,6 +477,7 @@ The following persistent volume types support mount options:
 - iSCSI
 - Azure Disk
 - Azure File
+- vSphere Volume
```
Comment (gnufied) [inline architecture/additional_concepts/storage.adoc]: We should call this VMWare Vsphere

File: install_config/persistent_storage/persistent_storage_vsphere_volume.adoc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+[[install-config-persistent-storage-persistent-storage-vsphere-volume]]
+= Persistent Storage Using vSphere Virtual Volume Plug-ins
```
Comment (gnufied) [inline install_config/persistent_storage/persistent_storage_vsphere_volume.adoc]: Same as above - the title should read "Persistent Storage using VMware vSphere"

**Comment by gaurav-nelson** [comment]
@gnufied Can you please check the updates?

**Comment by ron1** [comment]
A vSphere/VMWare "Virtual Volume", i.e., VVol, is a new capability available since vSphere 6.0 [see "vSphere 6.0 New Features – What is VMware Virtual Volumes (VVols)?" http://www.vmwarearena.com/vsphere-6-0-what-is-vmware-virtual-volumes-vvols/]. It is my understanding that the Kubernetes vSphere Cloud Provider does not currently support Virtual Volumes. As a result, please be careful to avoid references to the term "Virtual Volume" in the OpenShift documentation.

**Comment by gaurav-nelson** [comment]
@gnufied Up for review :globe_with_meridians: 
I have included most of the information, please let me know if something is missing.

**Comment by gaurav-nelson** [comment]
@gnufied bump!

**Comment by ron1** [comment]
OpenShift documentation should either reference Kubernetes vSphere Cloud Provider documentation or include most of what is contained therein. In particular, the content in sections "vSphere Cloud Provider" (https://kubernetes.io/docs/getting-started-guides/vsphere/#vsphere-cloud-provider) and "Enabling vSphere Cloud Provider Steps 1-7" (https://kubernetes.io/docs/getting-started-guides/vsphere/#enable-vsphere-cloud-provider) should be completely covered in their entirety. Specifically, Steps 1-4 seem to be completely missing. Also confirm that OpenShift 3.6 does not support "vSphere Storage Policy Based Management for Containers orchestrated by OpenShift/Kubernetes" and does not support Datastore Clusters as both of these appear to be Kubernetes 1.7/1.8 features. OpenShift docs should also describe Dynamic Provisioning of vSphere Volume PVCs in more detail including an example. Just a few thoughts.

File: install_config/configuring_vsphere.adoc [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+{product-title} can be configured to access
+link:https://www.vmware.com/au/products/vsphere.html[VMWare vSphere] VMDK Volumes, including
+xref:../install_config/persistent_storage/persistent_storage_vsphere.adoc#install-config-persistent-storage-persistent-storage-vsphere[using VMWare vSphere VMDK Volumes as persistent storage] for application data. After VMDL Volume is configured
```
Comment (jhou1) [inline install_config/configuring_vsphere.adoc]: `After VMDL Volume is configured`: `VMDL` -> `VMDK`.

**Review by jhou1 — CHANGES_REQUESTED** [review]

File: install_config/configuring_vsphere.adoc [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    scsicontrollertype = pvscsi
+----
+<1> vCenter username for the vSphere cloud provider.
+<2> vCenter password for the spcified user.
+<3> IP Address or FQDN for the vCenter server.
+<4> (Optional) Port number for the vCenter server. (Defaults to port `443`)
+<5> Set to 1 if the vCenter uses a self-signed cert.
+<6> Name of the data center on which Node VMs are deployed.
+<7> Name of the datastore to use for provisioning volumes using the storage classes or dynamic provisioning. If datastore is located in storage folder or datastore is member of datastore cluster, specify the full datastore path. Verify that vSphere Cloud Provider user has Read Privilege set on the datastore cluster or storage folder to be able to find datastore.
+<8> The vCenter VM folder path, in which the node VMs are located.  It can be set to an empty path(`working-dir = ""`), if Node VMs are located in the root VM folder.
```
Comment (jhou1) [inline install_config/configuring_vsphere.adoc]: `working-dir` is optional

**Comment by vikram-redhat** [comment]
Thanks @jhou1 .

@bfallonf - Please review, merge and do a follow up for the fixes as required.

**Comment by ahardin-rh** [comment]
@bfallonf Once this merges and you do any follow-ups, in a separate PR can you please add a link to this content in the[ vSphere section of release notes](http://file.rdu.redhat.com/~ahardin/07282017/ocp-3-6-release-notes/release_notes/ocp_3_6_release_notes.html#ocp-36-vmware-vsphere-storage)?

**Comment by ahardin-rh** [comment]
FYI @gaurav-nelson and @vikram-redhat on the above note

**Comment by bfallonf** [comment]
Thank you @jhou1 . I'll merge this and fix those in a separate PR.

@ahardin-rh will do :)

**Comment by bfallonf** [comment]
[rev_history]
|xref:../install_config/configuring_vsphere.adoc#install-config-configuring-vsphere[Configuring for VMWare vSphere]
|Added the Configuring for VMWare vSphere topic.
%
|xref:../install_config/persistent_storage/persistent_storage_vsphere.adoc#install-config-persistent-storage-persistent-storage-vsphere[Configuring Persistent Storage -> Persistent Storage Using VMWare vSphere Volume]
|Added the Persistent Storage Using VMWare vSphere Volume topic.
%

File: install_config/configuring_vsphere.adoc [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+<9> (Optional) VM name of the Master Node. If this parameter is specified , `vsphere.conf` file on the worker node does not need vCenter credentials.
```
Comment (jhou1) [inline install_config/configuring_vsphere.adoc]: According to https://kubernetes.io/docs/getting-started-guides/vsphere/, `vm-name` is introduced in kubernetes 1.6.4. Our latest OCP has kubernetes `v1.6.1+5115d708d7`. So I think this parameter isn't supported in our product yet.

Comment (bfallonf) [inline install_config/configuring_vsphere.adoc]: Thank you @jhou1 . This PR was merged, so I created another: #4962 

Please let me know if there's anything more for this PR.

**Pull request merged**